src/main/java/dev/sigstore/KeylessVerifier.java

```diff
@@ -0,0 +1,179 @@
+            InvalidKeySpecException, NoSuchAlgorithmException {
+      var fulcioCert =
+          Resources.toByteArray(
+              Resources.getResource("dev/sigstore/tuf/production/fulcio_v1.crt.pem"));
+      var ctfePublicKey =
+          Resources.toByteArray(Resources.getResource("dev/sigstore/tuf/production/ctfe.pub"));
+      var rekorPublicKey =
+          Resources.toByteArray(Resources.getResource("dev/sigstore/tuf/production/rekor.pub"));
+      fulcioVerifier(FulcioVerifier.newFulcioVerifier(fulcioCert, ctfePublicKey));
+      rekorClient(RekorClient.builder().build(), RekorVerifier.newRekorVerifier(rekorPublicKey));
+      return this;
+    }
+  }
+
+  public KeylessVerificationResult verifyOnline(
+      byte[] artifactDigest, byte[] certChain, byte[] signature)
+      throws IOException, NoSuchAlgorithmException, InvalidKeyException, CertificateException,
+          SignatureException {
+    var signingCert = SigningCertificate.from(Certificates.fromPemChain(certChain));
+    var leafCert = signingCert.getLeafCertificate();
+
+    var checks = KeylessVerificationResult.builder();
+
+    try {
+      fulcioVerifier.verifyCertChain(signingCert);
+      checks.checkPassed(Check.FULCIO_CERT);
+    } catch (FulcioVerificationException ex) {
+      return checks.checkFailed(Check.FULCIO_CERT, ex.getMessage());
+    }
+
+    try {
+      fulcioVerifier.verifySct(signingCert);
+      checks.checkPassed(Check.FULCIO_CERT_SCT);
+    } catch (FulcioVerificationException ex) {
+      return checks.checkFailed(Check.FULCIO_CERT_SCT, ex.getMessage());
+    }
+
+    var hashedRekordRequest =
+        HashedRekordRequest.newHashedRekordRequest(
+            artifactDigest, Certificates.toPemBytes(leafCert), signature);
+    Optional<RekorEntry> rekorEntry;
+    try {
+      rekorEntry = rekorClient.getEntry(hashedRekordRequest);
+    } catch (IOException ex) {
+      return checks.checkFailed(Check.REKOR_ENTRY_EXISTS, "retrieval failure: " + ex.getMessage());
+    }
+    if (rekorEntry.isEmpty()) {
+      return checks.checkFailed(Check.REKOR_ENTRY_EXISTS, "entry not found");
+    }
+    checks.checkPassed(Check.REKOR_ENTRY_EXISTS);
+
+    if (!hashedRekordRequest
+        .toJsonPayload()
+        .equals(
+            new String(
+                Base64.getDecoder().decode(rekorEntry.get().getBody()), StandardCharsets.UTF_8))) {
+      // TODO: can this happen? That would be wild
+      checks.checkFailed(
+          Check.REKOR_ENTRY_CONTENTS,
+          "provided signing information did not match retrieved rekor entry");
+    }
+    checks.checkPassed(Check.REKOR_ENTRY_CONTENTS);
+
+    try {
+      rekorVerifier.verifyEntry(rekorEntry.get());
+      checks.checkPassed(Check.REKOR_ENTRY_SIGNED);
+    } catch (RekorVerificationException ex) {
+      return checks.checkFailed(Check.REKOR_ENTRY_SIGNED, ex.getMessage());
+    }
+
+    try {
+      rekorVerifier.verifyInclusionProof(rekorEntry.get());
```

Comment:
Zach was mentioning that an inclusion proof is not something that clients should do by default, and that it's more for auditors of the log to make sure it's trustworthy.. Which reminds me we want a default verification method that doesn't hit Rekor.

Reply:
Yeah there would be a verifyOnline and verifyOffline (we probably need better naming too)

Reply:
I guess the question (probably a dumb one) I have is what is online verification for? What is the usecase?

Reply:
I would assume a package repository would want to do it at upload time? If it's only useful during log verification and completely unnecessary for individual artifacts -- maybe there's no point.

Kinda waiting on the bundle offline spec to finalize before anything. I think doing this now doesn't preclude us from changing it later (we would be 0.1.0 for people to play with). We can save the harder decisions for when we start to get some feedback?

Reply:
SG